pkg/kubelet/util/swap/swap_util.go

```diff
@@ -0,0 +1,109 @@
+package swap
+
+import (
+	"bytes"
+	"fmt"
+	"k8s.io/klog/v2"
+	"k8s.io/mount-utils"
+	"os"
+	sysruntime "runtime"
+	"strings"
+	"sync"
+)
+
+var (
+	tmpfsNoswapOptionSupported        bool
+	tmpfsNoswapOptionAvailabilityOnce sync.Once
+	swapOn                            bool
+	swapOnErr                         error
+	swapOnOnce                        sync.Once
+	rootDir                           string
+)
+
+const TmpfsNoswapOption = "noswap"
+
+// InitializeRootDir gets a path for a base directory, to which we have write permission.
+// This path will be later use for writing temporary files / mount points / etc.
+func InitializeRootDir(path string) {
+	rootDir = path
+}
+
+func IsTmpfsNoswapOptionSupported(mounter mount.Interface) bool {
+	isTmpfsNoswapOptionSupportedHelper := func() bool {
+		if sysruntime.GOOS == "windows" {
+			return false
+		}
+
+		baseMountDir := rootDir
+		if baseMountDir == "" {
+			klog.ErrorS(fmt.Errorf("swap utils root dir is not initialized, falling back to /tmp"), "")
+		}
+
+		mountDir, err := os.MkdirTemp(baseMountDir, "tmpfs-noswap-test-")
+		if err != nil {
+			klog.InfoS("error creating dir to test if tmpfs noswap is enabled. Assuming not supported", "mount path", mountDir, "error", err)
+			return false
+		}
+
+		defer func() {
+			err = os.RemoveAll(mountDir)
+			if err != nil {
+				klog.ErrorS(err, "error removing test tmpfs dir", "mount path", mountDir)
+			}
+		}()
+
+		err = mounter.MountSensitiveWithoutSystemd("tmpfs", mountDir, "tmpfs", []string{TmpfsNoswapOption}, nil)
+		if err != nil {
+			klog.InfoS("error mounting tmpfs with the noswap option. Assuming not supported", "error", err)
+			return false
+		}
+
+		err = mounter.Unmount(mountDir)
+		if err != nil {
+			klog.ErrorS(err, "error unmounting test tmpfs dir", "mount path", mountDir)
+		}
+
+		return true
+	}
+
+	tmpfsNoswapOptionAvailabilityOnce.Do(func() {
+		tmpfsNoswapOptionSupported = isTmpfsNoswapOptionSupportedHelper()
+	})
+
+	return tmpfsNoswapOptionSupported
+}
+
+// IsSwapOn detects whether swap in enabled on the system by inspecting
+// /proc/swaps. If the file does not exist, an os.NotFound error will be returned.
+// If running on windows, swap is assumed to always be false.
+func IsSwapOn() (bool, error) {
+	isSwapOnHelper := func() (bool, error) {
+		if sysruntime.GOOS == "windows" {
+			return false, nil
+		}
+
+		const swapFilePath = "/proc/swaps"
+		swapData, err := os.ReadFile(swapFilePath)
+		if err != nil {
+			if os.IsNotExist(err) {
+				klog.InfoS("File does not exist, assuming that swap is disabled", "path", swapFilePath)
```

Comment:
The procfs'  `/proc/swaps` file will always be there as long as `CONFIG_PROC_FS` is enabled, and this is usually true. There might be an issue reading the file, but that does not mean swap is disabled.

Reply:
> There might be an issue reading the file, but that does not mean swap is disabled.

I agree. The question is what we should do in such a scenario.
I've followed the previous implementation (before this PR), which assumed swap is not enabled if the file is not there.

Would you suggest to change this behavior? What do you have in mind?

Reply:
> > There might be an issue reading the file, but that does not mean swap is disabled.
> 
> I agree. The question is what we should do in such a scenario. I've followed the previous implementation (before this PR), which assumed swap is not enabled if the file is not there.

No worries. Makes sense to keep any prior art.
 
> Would you suggest to change this behavior? What do you have in mind?

This code in the kernel is almost 20 years old, and I have never seen this profs object not being there while procfs itself is enabled. Thus, I am not sure if checking for this file presence explicitly makes a lot of sense?

![Screenshot from 2024-05-09 17-47-01](https://github.com/kubernetes/kubernetes/assets/250723/142eb073-158b-4773-af79-8b55d47c9638)

However, it is probably best to leave this code as-is, if this was a lift from an existing implementation. :smile:

Reply:
:)
Let's defer this as well. I'm open for discussing removing it in the future :+1: